pkg/diff/sql_graph.go

```diff
@@ -4,64 +4,124 @@ import (
 	"fmt"
 
 	"github.com/stripe/pg-schema-diff/internal/graph"
-	"github.com/stripe/pg-schema-diff/internal/schema"
 )
 
+// sqlVertexId is an interface for a vertex id in the SQL graph
+type sqlVertexId interface {
+	fmt.Stringer
+}
+
+// sqlPriority is an enum for the priority of a statement in the SQL graph, i.e., whether it should be run sooner
+// or later in the topological sort of the graph
+type sqlPriority int
+
+const (
+	// Indicates a statement should run as soon as possible
+	sqlPrioritySooner sqlPriority = 1
+	// sqlPriorityUnset is the default priority for a statement
+	sqlPriorityUnset sqlPriority = 0
+	// Indicates a statement should run as late as possible
+	sqlPriorityLater sqlPriority = -1
+)
+
+// schemaObjSqlVertexId is a vertex id for a standard schema object node, i.e., indicating the creation/deletiion
+// of a schema object. It slots into the legacySqlVertexGenerator system.
+type schemaObjSqlVertexId struct {
+	objType  string
+	objId    string
+	diffType diffType
+}
+
+func buildSchemaObjVertexId(objType string, id string, diffType diffType) sqlVertexId {
+	return schemaObjSqlVertexId{
+		objType:  objType,
+		objId:    id,
+		diffType: diffType,
+	}
+}
+
+func (s schemaObjSqlVertexId) String() string {
+	return fmt.Sprintf("%s:%s:%s", s.objType, s.objId, s.diffType)
+}
+
 type sqlVertex struct {
-	ObjId      string
-	Statements []Statement
-	DiffType   diffType
+	// id is used to identify the sql vertex
+	id sqlVertexId
+
+	// priority is used to determine if the sql vertex should be included sooner or later in the topological
+	// sort of the graph
+	priority sqlPriority
+
+	// statements is the set of statements to run for this vertex
+	statements []Statement
 }
 
 func (s sqlVertex) GetId() string {
-	return fmt.Sprintf("%s_%s", s.DiffType, s.ObjId)
+	return s.id.String()
 }
 
-func buildTableVertexId(name schema.SchemaQualifiedName) string {
-	return fmt.Sprintf("table_%s", name)
+func (s sqlVertex) GetPriority() int {
+	// Prioritize adds/alters over deletes. Weight by number of statements. A 0 statement delete should be
+	// prioritized over a 1 statement delete
+	return len(s.statements) * int(s.priority)
 }
 
-func buildIndexVertexId(name schema.SchemaQualifiedName) string {
-	return fmt.Sprintf("index_%s", name)
+// dependency indicates an edge between the SQL to resolve a diff for a source schema object and the SQL to resolve
+// the diff of a target schema object
+//
+// Most SchemaObjects will have two nodes in the SQL graph: a node for delete SQL and a node for add/alter SQL.
+// These nodes will almost always be present in the sqlGraph even if the schema object is not being deleted (or added/altered).
+// If a node is present for a schema object where the "diffType" is NOT occurring, it will just be a no-op (no SQl statements)
+type dependency struct {
```

Comment:
nit: a doc on whether source depends on target or vice versa would be helpful

Reply:
oohh good call